ingestion-sink/src/main/java/com/mozilla/telemetry/ingestion/sink/config/SinkConfig.java

```diff
@@ -109,22 +148,67 @@ long getDefaultMaxOutstandingRequestBytes() {
+            env.getInt(BATCH_MAX_MESSAGES, 10_000), // default 10K messages
+            env.getDuration(BATCH_MAX_DELAY, "1s"), // default 1 second
+            PubsubMessageToTemplatedString.forBigQuery(env.getString(OUTPUT_TABLE)),
+            getFormat(env)));
+      }
+    },
+    bigQueryMixed {
+
+      @Override
+      Output getOutput(Env env) {
+        final com.google.cloud.bigquery.BigQuery bigQuery = getBigQueryService(env);
+        final Storage storage = getGcsService(env);
+        Function<PubsubMessage, CompletableFuture<Void>> bigQueryLoad = new BigQuery.Load(bigQuery,
+            storage,
+            // BigQuery Load API limits maximum bytes per request to 15TB
+            env.getLong(LOAD_MAX_BYTES, 15_000_000_000_000L), // default 15TB
+            // BigQuery Load API limits maximum files per request to 10,000
+            env.getInt(LOAD_MAX_FILES, 10_000), // default 10K files
+            // BigQuery Load API limits maximum loads per table per day to 1,000 but mixed mode
+            // expects fewer than 1,000 messages per day to need file loads
+            env.getDuration(LOAD_MAX_DELAY, "1s"), // default 1s
+            BigQuery.Load.Delete.always); // files will be recreated if not successfully loaded
+        // Combine bigQueryFiles and bigQueryLoad without an intermediate PubSub topic
+        Function<PubsubMessage, CompletableFuture<Void>> fileOutput = new Gcs.Write.Ndjson(storage,
+            env.getLong(BATCH_MAX_BYTES, 100_000_000L), // default 100MB
+            env.getInt(BATCH_MAX_MESSAGES, 1_000_000), // default 1M messages
+            env.getDuration(BATCH_MAX_DELAY, "10m"), // default 10 minutes
+            PubsubMessageToTemplatedString.forBigQuery(getBigQueryOutputBucket(env)),
+            getFormat(env),
+            blobInfo -> bigQueryLoad.apply(BlobInfoToPubsubMessage.apply(blobInfo)));
+        // Like bigQueryStreaming, but use STREAMING_ prefix for batch configuration
+        Function<PubsubMessage, CompletableFuture<Void>> streamingOutput = new BigQuery.Write(
+            bigQuery,
+            // BigQuery.Write.Batch.getByteSize reports protobuf size, which can be ~1/3rd more
+            // efficient than the JSON that actually gets sent over HTTP, so we use to 60% of the
+            // 10MB API limit by default.
+            env.getLong(STREAMING_BATCH_MAX_BYTES, 6_000_000L), // default 6MB
+            // BigQuery Streaming API Limits maximum rows per request to 10,000
+            env.getInt(STREAMING_BATCH_MAX_MESSAGES, 10_000), // default 10K messages
+            env.getDuration(STREAMING_BATCH_MAX_DELAY, "1s"), // default 1 second
+            PubsubMessageToTemplatedString.forBigQuery(env.getString(OUTPUT_TABLE)),
+            getFormat(env));
+        // BigQuery Streaming API Limits maximum row size to 1MiB
+        long maxStreamingSize = env.getLong(STREAMING_MESSAGE_MAX_BYTES, 1_000_000L); // default 1MB
```

Comment:
Note that Beam in a few places constructs default values in GiB/TiB like `(1L << 20)` for MiB. If we want to use exact values that match BQ limits, then I like that format. But `1_000_000L` is probably overall clearer to readers.

Reply:
I've been using the decimal values because it leaves room for disagreements about what constitutes size, but I also think it's easier to read than the bit-shift version.